File: android/src/main/kotlin/com/mytiki/sdk/capacitor/Receipt.kt
/*
 * Copyright (c) TIKI Inc.
 * MIT license. See LICENSE file in root directory.
 */

package com.mytiki.sdk.capacitor

import com.getcapacitor.JSArray
import com.getcapacitor.JSObject
import com.getcapacitor.PluginCall
import com.mytiki.tiki_sdk_android.channel.Channel
import com.mytiki.tiki_sdk_android.trail.TrailMethod
import com.mytiki.tiki_sdk_android.trail.req.ReqReceipt
import com.mytiki.tiki_sdk_android.trail.req.ReqReceiptAll
import com.mytiki.tiki_sdk_android.trail.req.ReqReceiptGet
import com.mytiki.tiki_sdk_android.trail.rsp.RspReceipt
import com.mytiki.tiki_sdk_android.trail.rsp.RspReceipts
import kotlinx.coroutines.MainScope
import kotlinx.coroutines.async

class Receipt(private val channel: Channel) {
    fun create(call: PluginCall) {
        val deferred = MainScope().async {
            val rsp: RspReceipt = channel.request(
                TrailMethod.RECEIPT_CREATE,
                ReqReceipt(
                    call.getString("payableId")!!,
                    call.getString("amount")!!,
                    call.getString("description"),
                    call.getString("reference")
                )
            ) { args ->
                RspReceipt.from(args)
            }.await()
            call.resolve(toJS(rsp))
        }
    }

    fun get(call: PluginCall) {
        val deferred = MainScope().async {
            val rsp: RspReceipt = channel.request(
                TrailMethod.RECEIPT_GET,
                ReqReceiptGet(call.getString("id")!!)
            ) { args ->
                RspReceipt.from(args)
            }.await()
            call.resolve(toJS(rsp))
        }
    }

    fun all(call: PluginCall) {
        val deferred = MainScope().async {
            val rsp: RspReceipts = channel.request(
                TrailMethod.RECEIPT_ALL,
                ReqReceiptAll(call.getString("payableId")!!)
            ) { args ->
                RspReceipts.from(args)
            }.await()
            val receipts: List<JSObject> =
                rsp.receipts?.map { receipt -> toJS(receipt) } ?: emptyList()
            val ret = JSObject()
            ret.put("receipts", JSArray(receipts))
            call.resolve(ret)
        }
    }

    companion object {
        fun toJS(receipt: RspReceipt): JSObject {
            val ret = JSObject()
            ret.put("id", receipt.id)
            ret.put(
                "payable",
                if (receipt.payable != null) Payable.toJS(receipt.payable!!) else null
            )
            ret.put("amount", receipt.amount)
            ret.put("description", receipt.description)
            ret.put("reference", receipt.reference)
            ret.put("timestamp", receipt.timestamp?.time)
            return ret;
        }
    }
}

File: ios/Plugin/Receipt.swift
/*
 * Copyright (c) TIKI Inc.
 * MIT license. See LICENSE file in root directory.
 */

import Foundation
import Capacitor
import TikiSdk

public class Receipt{
    
    public var tikiSdk = TikiSdk.config()
    
    @objc func create(_ call: CAPPluginCall) async {
        do{
            guard let payableId = call.getString("payableId"),
                  let amount = call.getString("amount") else{
                call.reject("Please provide payableId and amount to create a receipt.")
                return
            }
            let description = call.getString("description")
            let reference = call.getString("reference")

            let receipt = try await tikiSdk.trail.receipt.create(
                payableId: payableId,
                amount: amount,
                description: description,
                reference: reference
            )
            
            if(receipt != nil){
                call.resolve(Receipt.toJS(receipt!))
            }else{
                call.resolve([:])
            }
        }catch{
            call.reject(error.localizedDescription)
        }
    }
    
    @objc func get(_ call: CAPPluginCall) async {
        do{
            guard let id = call.getString("id") else {
                call.reject("Please provide the id of the Receipt")
                return
            }
            let receipt = try await tikiSdk.trail.receipt.get(
                id: id
            )
            if(receipt != nil){
                call.resolve(Receipt.toJS(receipt!))
            }else{
                call.resolve([:])
            }
        }catch{
            call.reject(error.localizedDescription)
        }
    }
    
    @objc func all(_ call: CAPPluginCall) async {
        do{
            guard let payableId = call.getString("payableId") else {
                call.reject("Please provide payableId, amount and type in plugin call")
                return
            }
            let receipt = try await tikiSdk.trail.receipt.all(
                payableId: payableId
            )
            call.resolve(["receipts": receipt.map{p in Receipt.toJS(p)}])
        }catch{
            call.reject(error.localizedDescription)
        }
    }
    
    static private func toJS(_ receipt: ReceiptRecord) -> JSObject{
        var ret = JSObject()
        ret["id"] = receipt.id
        ret["payable"] = Payable.toJS(receipt.payable)
        ret["amount"] = receipt.amount
        ret["description"] = receipt.description
        ret["reference"] = receipt.reference
        return ret
    }
}
